### PR TITLE
[WebAssembly] Upstream misc. EH changes

### PR DESCRIPTION
**Pull request by aheejin**
This upstreams more recent, mostly EH changes from libcxx and libcxxabi:
- `__cxa_init_primary_exception`-related changes made when updating to LLVM 18.1.2 (https://github.com/emscripten-core/emscripten/pull/21638)
- Removes ctype macros (https://github.com/emscripten-core/emscripten/pull/20960)
- Guard destructor changes with `__wasm__` (https://github.com/emscripten-core/emscripten/pull/21974)

### CODE REVIEW

**Comment by aheejin** [comment]
This fails clang-format, but the modified files don't conform to clang-format in the first place, and I tried to match the style the files were written in.

**Review by sbc100 — APPROVED** [review]

**Comment by ldionne** [comment]
As a matter of process, please do not merge patches to libc++/libc++abi/libunwind unless you have an approval from the relevant review group. Anyone is free to provide a code review, but the final approval must be granted by the review group. I know this wasn't done with a bad intention, I just want to make sure that our policies are understood and followed by folks.

On a different note, we should set up a CI job that builds and tests libc++ on Wasm. Our support policy mentions that we must have CI set up in order to officially support a platform, and what happened here is that patches of increasing intrusiveness were checked in and kinda got under the radar, so we never really discussed about the support policy for wasm. As it stands, wasm is technically not supported by libc++ so we'd be free to remove all the code as dead. Of course, we don't want to do that, but instead we should bring wasm in line with the other supported platforms and then we can officially claim support for it.

**Pull request merged**
